fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -0,0 +1,180 @@
+            this.compressed = false;
+            this.encrypted = false;
+        }
+
+        private void setCompressed(boolean compressed) {
+            this.compressed = compressed;
+        }
+
+        private void setEncrypted(boolean encrypted) {
+            this.encrypted = encrypted;
+        }
+
+        private boolean isCompressed() {
+            return compressed;
+        }
+
+        private boolean isEncrypted() {
+            return encrypted;
+        }
+    }
+
+    @Nullable
+    public static byte[] encode(@Nullable byte[] data, boolean compress, boolean encrypt) {
+        if (data == null) {
+            return null;
+        }
+
+        final EncodingState state = new EncodingState();
+        final ByteBuffersDataOutput decodedDataOutput = new ByteBuffersDataOutput();
+        // Placeholder for the code byte at beginning, will be modified in output byte array if needed
+        decodedDataOutput.writeByte((byte) 0);
+
+        byte[] encoded = compressIfNeeded(compress, data, decodedDataOutput, state, 1);
+
+        int code = 0;
+        if (state.isCompressed()) {
+            code = code | ENCODING_COMPRESSED;
+        }
+        // Encryption will be supported in future
+        if (state.isEncrypted()) {
+            code = code | ENCODING_ENCRYPTED;
+        }
+        encoded[0] = (byte) code;
+        LOGGER.trace("Encoded data for compressionSupposed={}, encryptionSupposed={}, compressedEventually={}, encryptedEventually={}, originalDataSize={}, encodedDataSize={}",
+                compress, encrypt, state.isCompressed(), state.isEncrypted(), data.length, encoded.length);
+        return encoded;
+    }
+
+    @Nullable
+    public static byte[] decode(@Nullable byte[] data) {
+        if (data == null) {
+            return null;
+        }
+
+        if (data.length < 2) {
+            throw new RecordCoreException("Invalid data")
+                    .addLogInfo(LogMessageKeys.DATA_VALUE, data);
+        }
+
+        final byte encoding = data[0];
+        final boolean encrypted = (encoding & ENCODING_ENCRYPTED) == ENCODING_ENCRYPTED;
+        final boolean compressed = (encoding & ENCODING_COMPRESSED) == ENCODING_COMPRESSED;
+
+        byte[] decoded = decompressIfNeeded(compressed, data, 1);
+        LOGGER.trace("Decoded data for compressed={}, encrypted={}, encodedDataSize={}, decodedDataSize={}",
+                compressed, encrypted, data.length, decoded.length);
+        return decoded;
+    }
+
+    @Nonnull
+    private static byte[] compressIfNeeded(boolean compressionNeeded, @Nonnull byte[] data, @Nonnull ByteBuffersDataOutput encodedDataOutput,
+                                           @Nonnull EncodingState state, int offset) {
+        if (!compressionNeeded) {
+            return fallBackToUncompressed(encodedDataOutput.toArrayCopy(), data, state, offset);
+        }
+
+        try (Compressor compressor = CompressionMode.HIGH_COMPRESSION.newCompressor()) {
```

Comment:
This is probably fine for now, but it's probably worth considering how we'd make this configurable in the future to allow for user control over high vs fast compression. Maybe this could become an Issue, so we don't lose track of it, or maybe not. Though for version 1, it being stuck on high compression seems fine

Reply:
Opened an issue: https://github.com/FoundationDB/fdb-record-layer/issues/1467